core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,304 @@
+ * @since 03-30-2018
+ */
+public class HeartBeatProvider implements HeartBeatProviderInterface, Stoppable {
+
+  /**
+   * The name of the heartbeat metric.
+   */
+  private final String HEARTBEAT_SYNTHETIC_METRIC_NAME = "HeartbeatState";
+
+  /**
+   * The list of disabled properties
+   */
+  private List<String> disableDefaultProperties = new ArrayList<>();
+
+  /**
+   * List of disabled heartbeat providers
+   */
+  private List<String> disabledHeartBeatPropertiesProviders = new ArrayList<>();
+
+  /**
+   * The counter for heartbeat sent to portal
+   */
+  private AtomicLong heartbeatsSent;
+
+  /**
+   * Map to hold heartbeat properties
+   */
+  private ConcurrentMap<String, HeartBeatPropertyPayload> heartbeatProperties;
+
+  /**
+   * Interval at which heartbeat would be sent
+   */
+  private long interval;
+
+  /**
+   * Telemetry client instance used to send heartbeat.
+   */
+  private TelemetryClient telemetryClient;
+
+  /**
+   * ThreadPool used for adding properties to concurrent dictionary
+   */
+  private ExecutorService propertyUpdateService;
+
+  /**
+   * Threadpool used to send data heartbeat telemetry
+   */
+  private ScheduledExecutorService heartBeatSenderService;
+
+  /**
+   * Heartbeat enabled state
+   */
+  private volatile boolean isEnabled;
+
+  private final Object lock = new Object();
+
+  public HeartBeatProvider() {
+    this.interval = HeartBeatProviderInterface.DEFAULT_HEARTBEAT_INTERVAL;
+    this.heartbeatProperties = new ConcurrentHashMap<>();
+    this.isEnabled = true;
+    this.heartbeatsSent = new AtomicLong(0);
+    this.propertyUpdateService = Executors.newCachedThreadPool(ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+    this.heartBeatSenderService = Executors.newScheduledThreadPool(1, ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+  }
+
+  @Override
+  public String getInstrumentationKey() {
+    return this.telemetryClient.getContext().getInstrumentationKey();
+  }
+
+  @Override
+  public void setInstrumentationKey(String key) {
+    if (this.telemetryClient != null && key != null) {
+      this.telemetryClient.getContext().setInstrumentationKey(key);
+    }
+  }
+
+  @Override
+  public void initialize(TelemetryConfiguration configuration) {
+    if (isEnabled) {
+      if (this.telemetryClient == null) {
+        this.telemetryClient = new TelemetryClient(configuration);
+      }
+
+      //Submit task to set properties to dictionary using separate thread. we do not wait for the
+      //results to come out as some I/O bound properties may take time.
+      propertyUpdateService.submit(HeartbeatDefaultPayload.populateDefaultPayload(getExcludedHeartBeatProperties(),
+          getExcludedHeartBeatPropertyProviders(), this));
+
+      heartBeatSenderService.scheduleAtFixedRate(heartBeatPulse(), interval, interval, TimeUnit.SECONDS);
+    }
+  }
+
+  @Override
+  public boolean addHeartBeatProperty(String propertyName, String propertyValue,
```

Comment:
I must say, I have strong feelings against this add/set property pattern. I don't understand why we need to throw if someone is overwriting a property with `add` or setting a new property with `set`. Please explain why we are doing it this way.

Reply:
@littleaj I believe it adds more readability and easier to distinguish operations. I know that it is not common in Java's map class however at the same time I also do feel that being some what opinionated here is not causing any blockage. Unless there is a mistake in documentation I don't think user will play around with this.

Reply:
One knob is simpler than two knobs; and there's no reason to be this strict. That's all I'm saying.

Reply:
@littleaj fair enough. I was under impression to change the api to have only set() method instead of add() and set() both. But from the comment now I understand that the only ask here is to not throw the exception but log the message accordingly which is acceptable. I can make this changes too.